snow/engine/snowman/ancestor/tree.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package ancestor
+
+import (
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+var _ Tree = (*tree)(nil)
+
+type Tree interface {
+	// Add a mapping from blkID to parentID.
+	//
+	// Invariant: blkID must not be equal to parentID
```

Comment:
It is expected that the `blkID` is the cryptographic hash of the `block`. Additionally, we expect the `parentID` to be included in the `block`. So these invariants should be maintained by these core assumptions of a blockchain... The only reason I'm explicitly documenting them is because of some fuzzing work that was done that found some really weird behavior here.